### ISSUE TITLE
Upload order invoice, feeds api, UPLOAD_VAT_INVOICE

### ISSUE THREAD

**Issue by magik092** [amzn/selling-partner-api-docs#404]
We are trying to upload order invoice using feedsApi (UPLOAD_VAT_INVOICE). 
Amazon ticket ID: 6780264492
Here is our scenario:

We are using authorization from refresh_token generated in seller central where we have published application. In this account we have those order.
**1) We are creating Feed Document:**
```
POST /feeds/2020-09-04/documents
{
    "contentType":"application/pdf"
}
```
Request passes, in response we can see resource URI.
**2) On the given URI, now we uploading encoded invoice file:**
```
PUT https://tortuga-prod-eu.s3-eu-west-1.amazonaws.com/%2FNinetyDays/amzn1.tortuga.3.deb962b5-c2 (...)
(encoded invoice file content, AES256 with key and iv from Feed Document)
```
At this point response is correct again (response code is 200).
**3) In last step we are creating Feed:**

```
POST /feeds/2020-09-04/feeds
{
    "feedOptions": {
        "DocumentType": "Invoice",
        "InvoiceNumber": "FV-EUR/1/2/3",
        "OrderId": "302-1 (...) 4",
        "TotalAmount": "39.3700",
        "TotalVATAmount": 6.29
    },
    "feedType": "UPLOAD_VAT_INVOICE",
    "inputFeedDocumentId": "amzn1.tortuga.3.deb962b5-c2 (...)",
    "marketplaceIds": [
        "A1PA6795UKMFR9"
    ]
}
```
In response there is 403 error:
```
[403] Client error: `POST https://sellingpartnerapi-eu.amazon.com/feeds/2020-09-04/feeds` resulted in a `403 Forbidden` response:
{
  "errors": [
    {
      "code": "Unauthorized",
      "message": "Access to the resource is forbidden",
      "details": ""
    }
  ]
}

#responseHeaders:
Date: Tue, 09 Feb 2021 07:53:02 GMT
x-amzn-RequestId: 44252153-0c4d- (...)
x-amz-apigw-id: ad7WzE (...)
X-Amzn-Trace-Id: Root=1-60223f5e-089 (...)
```

It looks like we have not enough some privileges but in seller central, in Develop Apps module in our hybrid (and published) app we have selected all required privileges. All the necessary steps in [selling partner api docs](https://github.com/amzn/selling-partner-api-docs/blob/main/guides/developer-guide/SellingPartnerApiDeveloperGuide.md#registering-your-selling-partner-api-application) have been completed.

A similar scenario, sending tracking information (using Feed Api) passes in 100% correctly.

![Clipboard - 11 lutego 2021 13_20](https://user-images.githubusercontent.com/7090340/107641596-1b012c80-6c74-11eb-8007-54193dfbbda7.png)

**Comment by fernlop** [comment]
Same here, I'm trying to get help with this since 10.12.20! 
All other Feed Types work, but this one returns a "Access to the resource is forbidden" response. 
At some point I got a response, that you need to get approved to upload invoices through the API. 

> Thus, if you require this feed type, you should contact the Amazon Business Seller program team (B2B) through upload-your-invoices@amazon.com (page 24, Contact Us) for further clarification as this method is owned by them, and they published the following document:
> https://m.media-amazon.com/images/G/01/B2B/DeveloperGuide/vat_calculation_service__dev_guide_H383rf73k4hsu1TYRH139kk134yzs.pdf

But they won't answer you too.

**Comment by LuigiPiccoli17** [comment]
Same thing here, could not find any documentation regarding the upload of invoices with the new SP-API

**Comment by magik092** [comment]
I have great news for you - UPLOAD_VAT_INVOICE has been implemented. We was already tested this scenario and we can confirm: it works!

**Comment by linzihao1245** [comment]
I have the same problem with you. magik092, How did you solve it? thank you !

**Comment by magik092** [comment]
@linzihao1245 if you follow this workflow (my first post) it now should work fine. Before my last comment, UPLOAD_VAT_INVOICE has not been implemented by Amazon Devs yet.

**Comment by linzihao1245** [comment]
> @ linzihao1245如果您遵循此工作流程（我的第一篇文章），它现在应该可以正常工作。在我最后发表评论之前，Amazon Devs尚未实现UPLOAD_VAT_INVOICE。

Yep,in step1.  I send a POST request to /feeds/2020-09-04/documents with contentType="application/pdf" , response 200 ,and got url、key and initializationVector.
Then , I send a PUT request with this [doc:](https://github.com/amzn/selling-partner-api-docs/blob/7bfacb95fc9ca2b3335a79acd40846deb3e9a161/guides/use-case-guides/feed-api-upload-invoice-guide-2020-09-04.md) . response 200 with succeed message.
Until step 3, send a POST request to /feeds/2020-09-04/feeds , I got 403 , unauthorized , Access to the resource is forbidden. such a trouble

**Comment by diegocvazquez** [comment]
@linzihao1245 
did you still get the 403 unauthorized in step 3? 
thanks

**Comment by diegocvazquez** [comment]
> Same here, I'm trying to get help with this since 10.12.20!
> All other Feed Types work, but this one returns a "Access to the resource is forbidden" response.
> At some point I got a response, that you need to get approved to upload invoices through the API.
> 
> > Thus, if you require this feed type, you should contact the Amazon Business Seller program team (B2B) through [upload-your-invoices@amazon.com](mailto:upload-your-invoices@amazon.com) (page 24, Contact Us) for further clarification as this method is owned by them, and they published the following document:
> > https://m.media-amazon.com/images/G/01/B2B/DeveloperGuide/vat_calculation_service__dev_guide_H383rf73k4hsu1TYRH139kk134yzs.pdf
> 
> But they won't answer you too.

@fernlop 

Hi, do you make any progress on this? I am also stucked at the 403 Unauthorized error, did you get any response from the B2B team? 

thanks!

**Comment by diegocvazquez** [comment]
I have tested this in MWS and it works, did someone make any progress regarding the 403 error in step 3?

**Comment by magik092** [comment]
@diegocvazquez did you try to contact with Amazon Team via Seller Central ticket? You can try to create ticket with all data needed (full requests, full responses). You have to be extremely patient but it can be best solution.

**Comment by wangjue666** [comment]
Maybe you can try setting **content type** to **"application/octet-stream"**  I haven't used the spaapi to dock here yet . But this method is effective in MWS :smile:

**Comment by diegocvazquez** [comment]
@wangjue666 
sorry I didnt mention that I have this working in MWS, but the issue is with SP-API, I have created a ticket still wating for an answer

**Comment by l-o-u-k-e** [comment]
@diegocvazquez, I'm in the same trouble, but with one difference: I cannot access MWS services because I signed up a couple of months ago. SP-API services are the only option to me only. Did they answer to your ticket?

**Comment by diegocvazquez** [comment]
@l-o-u-k-e 
Hi, yes I got an answer, the issue seems to be according to support that you need a tax role in oeder to be able to upload invoices via SP-API

**Comment by AnjaEk** [comment]
Hi @diegocvazquez, @magik092 
I know it is a while ago you were working on this issue, but we are having similar problems. Is it possible to share, how you wrapped up the file (UPLOAD_VAT_INVOICE) so that it was then successfully uploaded to Amazon? Did you went with xsd/xml or something else? Thank you in advance.

**Comment by diegocvazquez** [comment]
@AnjaEk 

Hi there, unfortunately I did not succeed on uploading the invoice. The process of uploading the file works okay, but when I try to create the feed I was getting an error, on further inquiring I have created a case, and the response was that the feed_type UPLOAD_VAT_INVOICE was not supported in SP-API as you can see in my comment. 

https://github.com/amzn/selling-partner-api-docs/issues/404#issuecomment-857726999

Actually some people posted here that there were able to achieve this, we left this because our developer profile does not include the Tax invoicing profile, and I was told that this is also related to the issue, so basically without this profiles there are certain endpoints related with invoicing that you cannot use you can see the profiles in the original post.
good luck

**Comment by davhol86** [comment]
Hi @magik092 

can u give more detailed info about Step 2 - Upload Feed Document of youre first post?
How did you strutured the body, which header parameters and authorization parameters did u use?

**Comment by magik092** [comment]
@davhol86

If you have the URL from step 1, you need to prepare the document content encrypted in AES and send using the PUT method without any authorization headers.
I have working code here: https://github.com/web-com-pl/amazon-selling-partner-api-php/blob/016b2d7f2bca710bd950b44c953d1cfb92662260/src/ResourcesApi/Api/ResourcesApi.php#L49

**Comment by davhol86** [comment]
@magik092 

Thank you, the first part is clear. You used the presigned url from the first step along with no authorization param and and the put Method. 
But my plan is to test this in postman and than implement it in C/AL (Microsoft Dynamics NAV).
For that i need to know how to structure the request json body, which I can't see in you're code.

 Another thing i need to know is if it's obligatory to encrypt it in AES or it's enough to convert it to base64?

**Comment by magik092** [comment]
@davhol86 
As far as I remember, AES is mandatory. The body is a raw string (encrypted document), not JSON.
eg.
`curl (...) --data-binary @'some-document.pdf'`

Here it is described on the documentation page how to perform an upload:
https://developer-docs.amazon.com/sp-api/docs/feeds-api-v2021-06-30-use-case-guide#step-3-upload-the-feed-data

**Comment by davhol86** [comment]
@magik092 
Now I've uploadet the file with postman. That worked, but now i get the same error as you in the first post.

`{
  "errors": [
    {
      "code": "Unauthorized",
      "message": "Access to the resource is forbidden",
      "details": ""
    }
  ]
}`

Strange. I used the JSON Body from your example, just changed the values...

Maybe cause i haven't encrypted the file in AES? I just selected the pdf from Postman.
![image](https://user-images.githubusercontent.com/11458257/169294364-2b2c0ea3-937a-4640-a179-d6156025269a.png)

I will try to encrypt the file and let you know.
